### PR TITLE
new API (in skikoMain): onClick and onDrag with PointerMatcher

### PR DESCRIPTION
**Pull request by eymar**
Added (in skikoMain):

Events handler api:
- fun Modifier.onClick
- suspend fun PointerInputScope.detectTapGestures
- fun Modifier.onDrag
- suspend fun PointerInputScope.detectDragGestures

- new class PointerInputMatcher - lets configure a condition which PointerEvnet should match to start/proceed/end a gesture
- new class PointerButton (PoinerEvent has a new property `val button:  PointerButton?`)

Notes/Questions (for the future):

1) Do we need `onPress` callback along with `onClick`? - It might be useful to select a component, but usually onClick and onPress show no noticeable difference.
2) Same for `onDoublePress`. Do we need it along with `onDoubleClick`?. Some apps use onDoublePress on some OS (e.g. windows). But usually the difference is not noticeable.
3) When we implement higher level clickable APIs such as onPrimaryClick or onContextMenuClick (secondary click), we should reuse onClick API added in this PR + add the rest of necessary Modifiers such as `indication`, `hoverable`, `onKey(Enter)`, etc.

### CODE REVIEW

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerButton.skiko.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import androidx.compose.ui.ExperimentalComposeUiApi
+import kotlin.jvm.JvmInline
+
+@JvmInline
+@ExperimentalComposeUiApi
+value class PointerButton(val index: Int) {
+    companion object {
+        val Primary = PointerButton(0)
+        val Secondary = PointerButton(2)
```
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerButton.skiko.kt]: In Compose, Secondary button has index=1 (different from Swing)

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerEvent.skiko.kt [hunk 2]
```diff
@@ -115,7 +121,9 @@ actual data class PointerEvent internal constructor(
      * - there was a synthetic move event sent by compose on relayout
      * - there was a synthetic move event sent by compose when move is missing between two non-move events
      */
-    val nativeEvent: Any?
+    val nativeEvent: Any?,
+
+    val button: PointerButton?
```
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerEvent.skiko.kt]: Better to mark it as ExperimentalComposeUiApi

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * Their combination is supported using plus operator:
+ * ```
+ * mouse(PointerButton.Primary) + touch + stylus + eraser
+ * ```
+ * See [DefaultPointerInputMatcher].
+ */
+@ExperimentalFoundationApi
+@OptIn(ExperimentalComposeUiApi::class)
+sealed interface PointerInputMatcher {
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: What if make it just `interface` and rename it to `PointerMatcher`?

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/gestures/DragGesture.skiko.kt [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+ * It follows the touch slop detection of [awaitTouchSlopOrCancellation]
+ * but will consume the position change automatically once the touch slop has been crossed.
+ * [onDragStart] will be called when touch slop in passed with the last known pointer position provided.
+ * [onDragEnd] is called after pointer matching [pointerInputMatcher] gets released,
+ * and [onDragCancel] is called if another gesture has consumed pointer input, canceling this gesture.
+ */
+@ExperimentalFoundationApi
+fun Modifier.onDrag(
+    enabled: Boolean = true,
+    pointerInputMatcher: PointerInputMatcher = PointerInputMatcher.DefaultPointerInputMatcher,
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/gestures/DragGesture.skiko.kt]: What if name the argument as `matcher`. It is more cleaner to use:
```
onDrag(
    matcher = PointerMatcher.mouse(PointerButton.Secondaty)
)
```

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        /**
+         * The Default [PointerInputMatcher] which covers the most common cases of pointer inputs.
+         * [DefaultPointerInputMatcher] will match [PointerEvent]s, which match at least one of the following conditions:
+         * - [PointerType] is [PointerType.Mouse] and [PointerEvent.button] is [PointerButton.Primary]
+         * - [PointerType] is [PointerType.Touch]
+         * - [PointerType] is [PointerType.Stylus], regardless of any buttons pressed
+         * - [PointerType] is [PointerType.Eraser]
+         */
+        @ExperimentalFoundationApi
+        val DefaultPointerInputMatcher = mouse(PointerButton.Primary) + touch + stylus + eraser
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: What if we name it as Primary? PointerMatcher.Primary

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt (hunk 3 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: rename to `PointerMatcher ` - Ack

make it just `interface`. 
Pros:
- more flexibility for users (they can override `fun matches(event: PointerEvent): Boolean`)
Cons:
- It will let users provide their own implementations (might be good for them, but...) and we will need to be more careful with this interface in the future. 

if we let users implement it, then we probably need to explain how it's used, so they provide correct implementations. 

for example, one can have such an implementation:

```kotlin
val matchAll = object : PointerMatcher {
   
   override val pointerType: PointerType = PointerType.Unknown
   
   override fun matches(event: PointerEvent) = true
}
```
It will probably work in the most common cases for onClick and onDrag, but might cause some "surprises".

I would wait for potential use cases when such custom implementations are needed. 

Maybe I'm overthinking on this :)

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerButton.skiko.kt (hunk 1 above)
Comment (eymar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerButton.skiko.kt]: fixed.
But it's good for me to know where are these indexes defined for Compose?

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt (hunk 3 above)
Comment (igordmn) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: It is a valid concern, but we have to explain this class anyway, what does it do (matches/filters all received press/release events, and reacts only on the events, which pass the match), so it can be safely used in user gesture detectors (triple click, for example). Without knowing that, users can't know where they should call `match`.

Another option, is to make `match` internal, but to much `internal` makes API too opinionated/restricted. And it can 
accidentality affect other future public API's.

We can provide antipatterns, how the user shouldn't override this class. For example, if we check key modifiers in this class, then we can miss the release event (check ctrl -> press ctrl -> press mouse -> release ctrl -> release mouse).

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt (hunk 5 above)
Comment (igordmn) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: Why not just `Primary` instead of `PrimaryMatcher`? `PointerMathcer.PrimaryMatcher` looks cumbersome in compare to `PointerMatcher.Primary` and this style isn't usually used in Compose (`Offset.Zero` instead of `Offset.ZeroOffset`)

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerButton.skiko.kt (hunk 1 above)
Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerButton.skiko.kt]: Look for `fun PointerButtons.isPressed(buttonIndex: Int)` function

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+ * mouse(PointerButton.Primary) + touch + stylus + eraser
+ * ```
+ * See [DefaultPointerInputMatcher].
+ */
+@ExperimentalFoundationApi
+@OptIn(ExperimentalComposeUiApi::class)
+sealed interface PointerInputMatcher {
+
+    @ExperimentalFoundationApi
+    val pointerType: PointerType
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: Can we move `pointerType` out of this interface, and make it implementation details of a subclass?

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt (hunk 3 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: Good!
made it an interface with `fun match(event: PointerEvent): Boolean`, moved `val pointerType: PointerType` to another private interface inheriting PointerMatcher.

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt (hunk 5 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: It was my mistake! The intention was to rename it to `Primary`

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt (hunk 6 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerInputMatcher.kt]: yes, It will be better. fixed.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt [hunk 7]
```diff
@@ -422,6 +423,7 @@ class ComposeScene internal constructor(
         buttons: PointerButtons? = null,
         keyboardModifiers: PointerKeyboardModifiers? = null,
         nativeEvent: Any? = null,
+        button: PointerButton? = null
```
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:428]: Please, add docs for this parameter (`@param button`) here and in `ImageComposeScene`

File: compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/window/WindowStateTest.kt [hunk 8]
```diff
@@ -27,11 +27,15 @@ import androidx.compose.ui.Alignment
 import androidx.compose.ui.ExperimentalComposeUiApi
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.awt.ComposeWindow
+import androidx.compose.ui.input.pointer.PointerKeyboardModifiers
```
Comment (igordmn) [inline compose/ui/ui/src/desktopTest/kotlin/androidx/compose/ui/window/window/WindowStateTest.kt]: Probably unnecessary changes (unused imports)

File: compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/platform/WindowInfo.kt [hunk 9]
```diff
@@ -22,7 +22,6 @@ import androidx.compose.runtime.Stable
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.rememberUpdatedState
 import androidx.compose.runtime.snapshotFlow
-import kotlinx.coroutines.flow.collect
```
Comment (igordmn) [inline compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/platform/WindowInfo.kt:25]: This import probably is needed in this file

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            override fun matches(event: PointerEvent): Boolean {
+                return event.changes.fastAll { it.type == pointerType }
+            }
+        }
+
+        private interface PointerTypeAndButtonMatcher : PointerTypeMatcher {
+            val button: PointerButton?
+
+            override fun matches(event: PointerEvent): Boolean {
+                return event.changes.fastAll { it.type == pointerType } &&
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt]: We could write `super.matches(event) &&`

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerButton.skiko.kt [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+package androidx.compose.ui.input.pointer
+
+import androidx.compose.ui.ExperimentalComposeUiApi
+import kotlin.jvm.JvmInline
+
+@JvmInline
+@ExperimentalComposeUiApi
+value class PointerButton(val index: Int) {
```
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerButton.skiko.kt:28]: Please, add docs for this class

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerEvent.skiko.kt [hunk 12]
```diff
@@ -136,9 +145,17 @@ actual data class PointerEvent internal constructor(
...
         keyboardModifiers = PointerKeyboardModifiers(0),
         _type = PointerEventType.Unknown,
-        nativeEvent = null
+        nativeEvent = null,
+        _button = null
     )
 
+    /**
+     * Specifies the pointer button state of which was changed.
+     * It has value only when event's type is [PointerEventType.Press] or [PointerEventType.Release].
```
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/pointer/PointerEvent.skiko.kt]: Worse to mention, that it applicable only to some pointers such as mouse/stylus, but not touch.

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+                    if (button != null) event.button == button else true
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt]: What if we need to differentiate a stylus press and a press of a stylus button? In this case this condition:
```
event.button == button
```
instead of
```
if (button != null) event.button == button else true
```
would make more sense

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+package androidx.compose.foundation
+
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.input.pointer.PointerButton
+import androidx.compose.ui.input.pointer.PointerEvent
+import androidx.compose.ui.input.pointer.PointerType
+import androidx.compose.ui.util.fastAll
+
+/**
+ * [PointerMatcher] represents a single condition or a set of conditions which a [PointerEvent] has to match
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt:26]: Worth to mention, that Compose for Desktop only supports `mouse` PointerType for now, and the other types are emulated as mouse.

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+    override suspend fun tryAwaitRelease(): Boolean {
+        if (!isReleased && !isCanceled) {
+            mutex.lock()
+        }
+        return isReleased
+    }
+}
+
+
+/*
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt]: Are these notes were accidentally added? :)

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+import androidx.compose.ui.util.fastAny
+import androidx.compose.ui.util.fastForEach
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+
+/**
+ * Configure component to receive clicks, double clicks and long clicks via input only (no accessibility "click" event)
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt:57]: Worth to mention, if the user wants to detect the primary click, better to use `clickable` modifier, because it adds indication, focusability, and Enter key handling

File: compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt [hunk 17]
```diff
@@ -0,0 +1,481 @@
...
+                        ) {
+                            clicksCount++
+                        }.size(10.dp, 20.dp)
+                )
+            }
+
+            scene.sendPointerEvent(PointerEventType.Move, Offset(0f, 0f))
+            scene.sendPointerEvent(PointerEventType.Press, Offset(0f, 0f), button = button)
+            scene.sendPointerEvent(PointerEventType.Release, Offset(0f, 0f), button = button)
+            delay(viewConfiguration.doubleTapTimeoutMillis * 2)
```
Comment (igordmn) [inline compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt:163]: Instead of delay it is better to use `UnconfinedTestDispatcher` (pass it to `ImageComposeScene`) with `dispatcher.scheduler.advanceTimeBy()`. It will be more robust and immune to any system performance degradation.

File: compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
+import org.junit.Ignore
+import org.junit.Test
+
+@OptIn(ExperimentalComposeUiApi::class, ExperimentalFoundationApi::class)
+class DragGestureTest {
+
+    @Test
+    fun `draggable by mouse primary button`() {
```
Comment (igordmn) [inline compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt:48]: It would be helpful, if we add a test for a drag with secondary button, where we check that it doesn't react on primary button.

File: compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+            }
+
+            // Note: touch slop is different from mouse slop,
+            // so values a bit different in this test comparing with mouse drag
+            scene.sendPointerEvent(PointerEventType.Move, Offset(5f, 5f), type = PointerType.Touch)
+            scene.sendPointerEvent(PointerEventType.Press, Offset(5f, 5f), type = PointerType.Touch)
+            scene.sendPointerEvent(
+                PointerEventType.Move,
+                Offset(5f + viewConfiguration.touchSlop + 5f, 5f),
+                type = PointerType.Touch
```
Comment (igordmn) [inline compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt:222]: Let's add a check that it doesn't react on mouse

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt [hunk 20]
```diff
@@ -0,0 +1,444 @@
...
+    enabled: Boolean = true,
+    interactionSource: MutableInteractionSource,
+    matcher: PointerMatcher = PointerMatcher.Primary,
+    keyboardModifiers: PointerKeyboardModifiers.() -> Boolean = { true },
+    onDoubleClick: (() -> Unit)? = null,
+    onLongClick: (() -> Unit)? = null,
+    onClick: () -> Unit
+) = composed(
+    inspectorInfo = {
+        name = "onCombinedClick"
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt]: The names should be "onClick" and "matcher" (not trigger)

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt [hunk 21]
```diff
@@ -0,0 +1,444 @@
...
+ * @param matcher defines supported pointer types and required properties
+ * @param keyboardModifiers defines a condition that [PointerEvent.keyboardModifiers] has to match
+ * @param onLongClick will be called when user long presses on the element
+ * @param onDoubleClick will be called when user double clicks on the element
+ * @param onClick will be called when user clicks on the element
+ */
+@ExperimentalFoundationApi
+fun Modifier.onClick(
+    enabled: Boolean = true,
+    interactionSource: MutableInteractionSource,
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt:124]: Similar to `combinedClickable`, we should handle the case when the composable is disposed (send a cancel in DisposableEffect)

File: compose/desktop/desktop/samples/src/jvmMain/kotlin/androidx/compose/desktop/examples/mouseclicks/Main.jvm.kt [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+
+                println("Recomposing now")
+
+                Column(modifier = Modifier.padding(25.dp)) {
+                    Text("onDrag")
+                    var offset1 by remember { mutableStateOf(Offset.Zero) }
+                    Box(modifier = Modifier.offset { IntOffset(offset1.x.toInt(), offset1.y.toInt()) }
+                        .size(100.dp).background(Color.Blue).pointerInput(Unit) {
+                            detectDragGestures(
+                                matcher = PointerMatcher.mouse(PointerButton.Secondary),
```
Comment (igordmn) [inline compose/desktop/desktop/samples/src/jvmMain/kotlin/androidx/compose/desktop/examples/mouseclicks/Main.jvm.kt:150]: When I perform a drag on Mac with two fingers on touchpad, then release one finger, I still perform a drag. Is we have a bug here, or mac continues to treat a single finger as the right button?

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt [hunk 23]
```diff
@@ -0,0 +1,444 @@
...
+
+    while (currentCoroutineContext().isActive) {
+        awaitPointerEventScope {
+            pressScope.reset()
+            while (
+                currentEvent != EmptyPointerEvent &&
+                currentEvent.isAllPressedDown(false) &&
+                matcher.matches(currentEvent)
+            ) {
+                // suspend until currentEvent matches the filter for pressed event
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt]: Do we need this code? When I debugged it, `isAllPressedDown` was always false

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/gestures/DragGesture.skiko.kt [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+ */
+@ExperimentalFoundationApi
+fun Modifier.onDrag(
+    enabled: Boolean = true,
+    matcher: PointerMatcher = PointerMatcher.Primary,
+    onDragStart: (Offset) -> Unit = {},
+    onDragCancel: () -> Unit = {},
+    onDragEnd: () -> Unit = {},
+    onDrag: (Offset) -> Unit
+): Modifier = composed {
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/gestures/DragGesture.skiko.kt]: Please, add `inspectorInfo`

**Review by igordmn — COMMENTED** [review]
API looks great, thanks!

Some comments regarding the implementation and the docs.

File: compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt (hunk 19 above)
Comment (eymar) [reply compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt:222]: I added a test to check that when onDrag expects touch events only, then drag will not start if the gesture starts using the mouse.

But, at the moment we can't have a test to check that onDrag(matcher=touch) never reacts on mouse, because `pointerInputEvent()` always creates a list of changes with hardcoded pointerId = 0. Therefore drag will keep running regardless of whether it's mouse or touch.  `drag(pointerId)` doesn't check the pointer type, but it only checks the pointerId, which makes sense.

But we'll add such a test on android when upstreaming this, or when we support touch/stylus inputs on desktop.

File: compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt (hunk 18 above)
Comment (eymar) [reply compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt:48]: Added a test to check that `onDrag with matcher = PointerMatcher.mouse(PointerButton.Secondary)` ignores events when Primary button performs press and slop., but start the drag when Secondary press and slop

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt (hunk 15 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt]: I wanted to keep them somewhere.  I remove them from the code and post it here just in case:

```
/*
 * Other notes:
 * 1) onPress? - might be useful to select a component, but usually onClick and onPress show no noticeable difference.
 * 2) onDoubleClick. It can be onDoublePress on some OS (e.g. windows). But usually the difference is not noticeable.
 * 3) no indication. users can add it along with interactionSource
 * 4) name = onClick
 *
 * clickable can use "onClick" + add indnication, + hoverable + focusable + onKey(enter) == Click
 * onContextMenuClick can use "onClick" too
 */
```

File: compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/platform/WindowInfo.kt (hunk 9 above)
Comment (eymar) [reply compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/platform/WindowInfo.kt:25]: `collect` is an extension function, but since coroutines 1.6.0 it's also a member function of Flow interface, so import is not needed (Ide highlights it is unused).

File: compose/desktop/desktop/samples/src/jvmMain/kotlin/androidx/compose/desktop/examples/mouseclicks/Main.jvm.kt (hunk 22 above)
Comment (eymar) [reply compose/desktop/desktop/samples/src/jvmMain/kotlin/androidx/compose/desktop/examples/mouseclicks/Main.jvm.kt:150]: Same on my Mac. I don't think it's a bug (at least not on our side), as we simply receive Secondary button as pressed even though only one finger interacts with TouchPad.

File: compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt (hunk 17 above)
Comment (eymar) [reply compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt:163]: This doesn't work in our case because UnconfinedDisptacher won't be used in pointerInput scope.
PointerInput scope context shows that it uses a different dispatcher -  `private val effectDispatcher = FlushCoroutineDispatcher(coroutineScope)`

and we don't control effectDispatcher from ImageComposeScene now. Should we make it possible to override it from testss?

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt (hunk 23 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt]: It was necessary when we relied on PointerButtons.
But now, we use PointerButton, so it's not needed.

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt (hunk 13 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt]: If someone wants to differentiate between a stylus press and stylus button press, then I think they need to just add 2 onClick modifiers in correct order:
```kotlin
onClick(matcher = PointerMatcher.stylus) {}
.onClick(matcher=PointerMatcher.stylus(PointerButton.Primary)) {}
```

If we had `event.button == button` instead of `if (button != null) event.button == button else true` then it means that button is required to be null. So we won't handle a simple stylus press with button pressed in cases when it's not important whether anything is pressed or not. Imagine taking a stylus away from the screen, then moving it to the screen and pressing on it. 


I think that `PointerMatcher.stylus` should mean that we don't really care about buttons, they can be pressed or not.
But I might be wrong.

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt [hunk 25]
```diff
@@ -0,0 +1,447 @@
...
+            val pressedInteraction = remember { mutableStateOf<PressInteraction.Press?>(null) }
+            val onClickState = rememberUpdatedState(onClick)
+            val on2xClickState = rememberUpdatedState(onDoubleClick)
+            val onLongClickState = rememberUpdatedState(onLongClick)
+            val keyboardModifiersState = rememberUpdatedState(keyboardModifiers)
+
+            val hasLongClick = onLongClick != null
+            val hasDoubleClick = onDoubleClick != null
+
+            DisposableEffect(hasLongClick) {
```
Comment (igordmn) [inline compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt:153]: If we have `PressedInteractionSourceDisposableEffect`, this effect isn't needed

File: compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt (hunk 19 above)
Comment (igordmn) [reply compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/gestures/DragGestureTest.kt:222]: > I added a test to check that when onDrag expects touch events only, then drag will not start if the gesture starts using the mouse.

Yes, I meant this test, not the test when we start a drag with touch, and continue the drag with mouse.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt (hunk 7 above)
Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:428]: And in `ImageComposeScene` :)

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt (hunk 15 above)
Comment (igordmn) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt]: Some of the points isn't understandable without the context. Maybe we can remove p.4 and clarify what onContextMenuClick means (or just remove it), because it isn't added yet. (and fix typo in indnication :) )

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt (hunk 13 above)
Comment (igordmn) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt]: > So we won't handle a simple stylus press with button pressed

But isn't the `button` is a change of the state? I.e. even if press with a stylus, we have `button` as null. If we want to handle the case when we need a press when the primary button is pressed, we need to add to matcher `buttons`, not `button`.

Here, if we write:
`onClick(matcher = PointerMatcher.stylus) {}
.onClick(matcher=PointerMatcher.stylus(PointerButton.Primary)) {}`, we will have two callbacks called when we press the primary button on the stylus, because in the first `onClick` all presses passes the filter.

File: compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt (hunk 17 above)
Comment (igordmn) [reply compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt:163]: `FlushCoroutineDispatcher` wraps the dispatcher passed into `ImageScene`. I.e. if it would be `UnconfinedTestDispatcher`, the functionality of `FlushCoroutineDispatcher` just won't be used, because unconfined dispatcher performs all callbacks immediately, and on flush there won't be any callbacks.

> and we don't control effectDispatcher from ImageComposeScene now.

It should be controlled, because `effectDispatcher` in fact will be `UnconfinedTestDispatcher`. Theoretically it should work. If you just set this dispatcher in `ImageComposeScene` constructor, will the test pass? If we it doesn't pass, then we have a bug in `ImageCompoScene`, but we can fix it later.

Comment (eymar) [reply compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt:163]: When I simply pass `UnconfinedTestDispatcher ` into `ImageComposeScene ` and try to advanceTime using that dispatcher in the test, the test fails. 

It seems, it has something to do with a fact that FlashCoroutineDispatcher doesn't implement a Delay needed in `withTimeoutOrNull`, so DefaultDelay is used - `Dispatchers.Default`, which doesn't know about UnconfinedTestDispatcher.

Comment (igordmn) [reply compose/foundation/foundation/src/desktopTest/kotlin/androidx/compose/foundation/OnClickTest.kt:163]: Oh, I understand. We should fix that someday, because it is also used in the usual Compose unit tests with `createComposeRule`. I will look onto this, when I will be fixing  the other ImageComposeScene bugs.

For now, let's just keep these tests as they are now.

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt (hunk 13 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt]: I see your point. 

> But isn't the `button` is a change of the state?

yes, it's a change of the state. So my example was incorrect. 

I agree, that `event.button == button` makes more sense now. I'll commit the change with your suggestion.

So when `matcher=PointerMatcher.stylus`, it   would mean that it's required that no button should be pressed on the stylus. it's equivalent to a click using touch.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt (hunk 7 above)
Comment (eymar) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:428]: fixed

File: compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt (hunk 15 above)
Comment (eymar) [reply compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/OnClick.skiko.kt]: I removed these notes from the code. Let's keep them in PR description.

**Comment by igordmn** [comment]
One small thing we need to do - deprecate [mouseClickable](https://github.com/JetBrains/compose-jb/tree/master/tutorials/Mouse_Events#mouse-rightmiddle-clicks-and-keyboard-modifiers), and rewrite the tutorial about it.

This can be done in another PR.

**Review by igordmn — APPROVED** [review]
